<a href="https://colab.research.google.com/github/btg1631/eCommerce-project/blob/main/docs/analysis/yohan/fifth_process.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install konlpy

In [ ]:
import warnings

# 경고 메시지를 무시하도록 설정
warnings.filterwarnings("ignore")

In [ ]:
import pandas as pd
from konlpy.tag import Okt

In [ ]:
# 모든 데이터 로드
df_bibim1 = pd.read_csv("./local.bibim1.csv")
df_bibim2 = pd.read_csv("./local.bibim2.csv")
df_bibim3 = pd.read_csv("./local.bibim3.csv")
df_bibim4 = pd.read_csv("./local.bibim4.csv")

df_oil1 = pd.read_csv("./local.oil1.csv")
df_oil2 = pd.read_csv("./local.oil2.csv")
df_oil3 = pd.read_csv("./local.oil3.csv")
df_oil4 = pd.read_csv("./local.oil4.csv")
df_oil5 = pd.read_csv("./local.oil5.csv")

df_ham1 = pd.read_csv("./local.ham1.csv")
df_ham2 = pd.read_csv("./local.ham2.csv")
df_ham3 = pd.read_csv("./local.ham3.csv")
df_ham4 = pd.read_csv("./local.ham4.csv")
df_ham5 = pd.read_csv("./local.ham5.csv")

df_kyoja1 = pd.read_csv("./local.kyoja1.csv")
df_kyoja2 = pd.read_csv("./local.kyoja2.csv")
df_kyoja3 = pd.read_csv("./local.kyoja3.csv")
df_kyoja4 = pd.read_csv("./local.kyoja4.csv")

df_tomato1 = pd.read_csv("./local.tomato1.csv")
df_tomato2 = pd.read_csv("./local.tomato2.csv")
df_tomato3 = pd.read_csv("./local.tomato3.csv")

df_canola1 = pd.read_csv("./ssg.canola_oil1.csv")
df_canola2 = pd.read_csv("./ssg.canola_oil2.csv")
df_canola3 = pd.read_csv("./ssg.canola_oil3.csv")
df_canola4 = pd.read_csv("./ssg.canola_oil4.csv")

df_kimchi1 = pd.read_csv("./ssg.kimchi1.csv")
df_kimchi2 = pd.read_csv("./ssg.kimchi2.csv")
df_kimchi3 = pd.read_csv("./ssg.kimchi3.csv")
df_kimchi4 = pd.read_csv("./ssg.kimchi4.csv")

df_mung_bean_sprouts1 = pd.read_csv("./ssg.mung_bean_sprouts1.csv")
df_mung_bean_sprouts2 = pd.read_csv("./ssg.mung_bean_sprouts2.csv")
df_mung_bean_sprouts3 = pd.read_csv("./ssg.mung_bean_sprouts3.csv")
df_mung_bean_sprouts4 = pd.read_csv("./ssg.mung_bean_sprouts4.csv")

df_pancake_powder1 = pd.read_csv("./ssg.pancake_powder1.csv")
df_pancake_powder2 = pd.read_csv("./ssg.pancake_powder2.csv")
df_pancake_powder3 = pd.read_csv("./ssg.pancake_powder3.csv")
df_pancake_powder4 = pd.read_csv("./ssg.pancake_powder4.csv")

df_rice1 = pd.read_csv("./ssg.rice1.csv")
df_rice2 = pd.read_csv("./ssg.rice2.csv")
df_rice3 = pd.read_csv("./ssg.rice3.csv")
df_rice4 = pd.read_csv("./ssg.rice4.csv")
df_rice5 = pd.read_csv("./ssg.rice5.csv")

df_bread1 = pd.read_csv("./local.bread_1.csv")
df_bread2 = pd.read_csv("./local.bread_2.csv")
df_bread3 = pd.read_csv("./local.bread_3.csv")
df_bread4 = pd.read_csv("./local.bread_4.csv")

df_cheese1 = pd.read_csv("./local.cheese_1.csv")
df_cheese2 = pd.read_csv("./local.cheese_2.csv")
df_cheese3 = pd.read_csv("./local.cheese_3.csv")

df_milk1 = pd.read_csv("./local.milk_1.csv")
df_milk2 = pd.read_csv("./local.milk_2.csv")
df_milk3 = pd.read_csv("./local.milk_3.csv")

df_water1 = pd.read_csv("./local.water_1.csv")
df_water2 = pd.read_csv("./local.water_2.csv")
df_water3 = pd.read_csv("./local.water_3.csv")

df_yogurt1 = pd.read_csv("./local.yogurt_1.csv")
df_yogurt2 = pd.read_csv("./local.yogurt_2.csv")
df_yogurt3 = pd.read_csv("./local.yogurt_3.csv")

dfs = [df_bibim1,df_bibim2,df_bibim3,df_bibim4,df_oil1,df_oil2,df_oil3,df_oil4,df_ham1,df_ham2,df_ham3,df_ham4,df_ham5,df_kyoja1,df_kyoja2,df_kyoja3,df_kyoja4,df_tomato1,df_tomato2,df_tomato3,df_canola1,df_canola2,df_canola3,df_canola4,df_kimchi1,df_kimchi2,df_kimchi3,df_kimchi4,df_mung_bean_sprouts1,df_mung_bean_sprouts2,df_mung_bean_sprouts3,df_mung_bean_sprouts4,df_pancake_powder1,df_pancake_powder2,df_pancake_powder3,df_pancake_powder4,df_rice1,df_rice2,df_rice3,df_rice4,df_rice5,df_bread1,df_bread2,df_bread3,df_bread4,df_cheese1,df_cheese2,df_cheese3,df_milk1,df_milk2,df_milk3,df_water1,df_water2,df_water3,df_yogurt1,df_yogurt2,df_yogurt3]

- 각 데이터에 대한 긍정 부정 비교 함수 제작

In [ ]:
# 3점을 제외한 나머지 라벨링 작업 함수 제작
def labeling(text):
    if text in [4,5]:
        text = 1
    elif text in [1,2]:
        text = 0
    else:
        pass
    return text

In [ ]:
df_bibim1[df_bibim1['textLevel'] == 1].shape[0]

3

In [ ]:
def compare_data(dfs_list,labeling_fuc):
    counter = 0
    save_data_list = []
    for i in dfs_list:
        # db저장시 발생했던 고유값 제거
        i = i.drop(columns = '_id')

        # 라벨링 함수 적용
        i['textLevel'] = i['textLevel'].apply(labeling_fuc)


        # 딥러닝이슈로 인한 중간값 제외
        processing_df = i.query('textLevel != 3')

        # 듀플 제거전 긍정/부정 갯수 카운트
        posi_count = processing_df[processing_df['textLevel'] == 1].shape[0]
        naga_count = processing_df[processing_df['textLevel'] == 0].shape[0]

        # 중복값 제거
        processing_df = processing_df.drop_duplicates()

        deldu_posi_count = processing_df[processing_df['textLevel'] == 1].shape[0]
        deldu_naga_count = processing_df[processing_df['textLevel'] == 0].shape[0]

        # print(f'{i["merchTitle"][0]} 제품의 제거전 긍정문의 갯수 : {posi_count}')
        # print(f'{i["merchTitle"][0]} 제품의 제거전 부정문의 갯수 : {naga_count}')
        # print(f'{i["merchTitle"][0]} 제품의 제거후 긍정문의 갯수 : {deldu_posi_count}')
        # print(f'{i["merchTitle"][0]} 제품의 제거후 긍정문의 갯수 : {deldu_naga_count}')

        # 2차원 리스트로 저장
        append_list = [posi_count,naga_count,deldu_posi_count,deldu_naga_count]
        save_data_list.append(append_list)

        counter += 1
        if counter != len(dfs_list):
            # print('')
            pass

    return save_data_list

In [ ]:
save_data_list = compare_data(dfs,labeling)

In [ ]:
rate_list = []
for i in save_data_list:
    rate = ((i[0] - i[2])/i[0])*100
    rate_list.append(rate)
    # print(rate)

- 스팸을 알아보기 위한 각 제품의 긍정문 토픽 추출

In [ ]:
# 기본적인 불용어와 상표명 등의 독립단어 설정
stopword1 = ['도', '는', '다', '의', '가', '이', '은', '한', '에', '하', '고', '을', '를', '인', '듯', '과', '와', '네', '들', '듯', '지', '임', '게','의','가','들','는','좀','잘','걍','과','도','를','으로','자','에','와','한','하다']
stopword2 = ['팔도','팔도비빔면','비빔면','괄도','네넴띤','괄도네넴띤','농심','배홍동','배홍동비빔면','오뚜기','진비빔면','삼양','찰비빔면','비비고','교자','왕교자','노브랜드','교자만두','만두','왕교자만두','피코크','담백한왕교자','토마토','청정원','파스타소스','폰타나','샘표','뽀모도로','CJ','cj','스팸','클래식','덴마크튤립햄','튤립햄','동원','리챔','하림','닭가슴살햄','챔','더본코리아','빽햄','백종원','고소한참기름','참기름','경천','경천식품','시골참기름','들기름','최순희','백설','통참깨참기름','통참깨','임금님','임금님표','이천쌀','한눈에 반한 쌀','경기미','농협','농협양곡','당진해나루쌀','삼광','의성','의성농협','바른고을','의성진쌀','김치','별미총각김치','총각김치','종가집','종가','오래오래맛있는총각김치','조선호텔특제육수','조선호텔','부침가루','부침','곰표','통감자','자연주의','숙주나물','친환경','국산숙주','숙주','무농약','통통숙주','카놀라유','카놀라오일','카놀라','Coosur','coosur','쿡투게더','부드러운우유식빵','식빵','우유식빵','삼립','롯데','롯데웰푸드','기린','우유','더 메나쥬리','메나쥬리','온가족우유식빵','삼다수','에비앙','백산수','플레인','요거트','덴마크','에치와이','hy','윌','상하','상하치즈','체다','치즈','서울우유','서울','슬라이스','남양','드빈치','체다치즈','갓밀크','굿모닝','굿밀크','헬로우','앙팡']
stopword = stopword1 + stopword2

In [ ]:
# Okt 형태소 분석기 인스턴스 생성
okt = Okt()

import re
import numpy as np


# 토크나이징 함수 정의
def tokenizer(raw, pos=["Noun","Alpha","Verb","Number"], stopword=stopword):

  get_list = []

  for word, tag in okt.pos(raw,norm=True,stem=True):
    if len(word) <= 1:
      word = '공란'

    if tag in pos and word not in stopword:
      get_list.append(word)
    else:
      get_list.append('공란')
    pass


  return get_list

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer = TfidfVectorizer(tokenizer=tokenizer, use_idf=True)

- LDA 토픽 추출 함수 제작

In [ ]:
def LDAs(vectorizer,input_text_list):


  features = vectorizer.fit_transform(input_text_list)

  # 단어 사전 추출
  dictionary_list = vectorizer.get_feature_names_out()

  # LDA 학습
  from sklearn.decomposition import LatentDirichletAllocation
  lda_model = LatentDirichletAllocation(n_components=5, random_state=55) #인스턴스화 #n_components 토픽의 갯수
  lda_model.fit(features) #교육

  ## 상위 단어 추출
  ## 0 확률 1은 dictionary
  topics_list = list()
  for topic in lda_model.components_:
      df_datas = [topic, dictionary_list]
      df_topics = pd.DataFrame(data=df_datas)
      df_topics= df_topics.T
      df_topics = df_topics.sort_values(0, ascending=False)

      topics_text = ' '.join(df_topics[1].values[:10])# 시리즈 형식으로 출력 get values from series / index
      topics_list.append(topics_text)

  return topics_list

In [ ]:
get_topics = []
counter = 0
for i in dfs:
  # db저장시 발생했던 고유값 제거
  i = i.drop(columns = '_id')

  # 라벨링 함수 적용
  i['textLevel'] = i['textLevel'].apply(labeling)

  # 긍정문 추출
  processing_df = i.query('textLevel == 1')

  # 토픽 추출
  try:
    get_topics.append(LDAs(vectorizer,processing_df['textContent']))
  except:
    get_topics.append(f'오류 발생 : {counter}번재 csv')

  counter += 1

In [ ]:
counter = 0

for i in get_topics:
  for j in i:
    if '공란' in j:
      counter += 1

print(f'stopwrd로 제거된 공랸 갯수 : {counter}')

stopwrd로 제거된 공랸 갯수 : 279


- 총 285(57개 csv * 5개 토픽주제) 중 279개에 공란이 들어있음
- stopword에서 상표같은 주제들과 단어 관사들만 제한 것으로 미루어 보았을 때, 상표명들만 단순히 적혀있는 댓글들이 상당수 있었던 것으로 예상할 수 있다.

In [ ]:
# csv 별 토픽 추출
total_counter = 0
for i in dfs:
  list_counter = 0
  for j in range(5):
    list_counter += 1
    print(f'{i["merchTitle"][0]} {list_counter}번째 토픽 : {get_topics[total_counter][j]}')

  total_counter += 1

  print('--------------------------------------')

팔도비빔면 130g*4개 1번째 토픽 : 공란 재다 먹다 비비다 아주 시키다 가끔 정말 잇다 두다
팔도비빔면 130g*4개 2번째 토픽 : 공란 배송 받다 감사하다 제품 항상 자다 주문 애용 상품
팔도비빔면 130g*4개 3번째 토픽 : 공란 가격 추천 품질 배송 대비 만족 골뱅이 유통 넣다
팔도비빔면 130g*4개 4번째 토픽 : 공란 구매 자주 역시 여름 최고 상품 제일 사다 제품
팔도비빔면 130g*4개 5번째 토픽 : 공란 먹다 자다 구입 아이 여름 가성 겨울 삼겹살 필수
--------------------------------------
농심 배홍동비빔면 4입 1번째 토픽 : 공란 배송 매콤 않다 먹다 시키다 소스 다른 갈아타다 조금
농심 배홍동비빔면 4입 2번째 토픽 : 공란 추천 잇다 배홍 먹기 먹다 동이 광고 재다 보고
농심 배홍동비빔면 4입 3번째 토픽 : 공란 자다 먹다 받다 배송 비비다 이용 보다 이제 요즘
농심 배홍동비빔면 4입 4번째 토픽 : 공란 배홍 제일 사다 처음 가격 먹다 아이 여름 구입
농심 배홍동비빔면 4입 5번째 토픽 : 공란 구매 자주 제품 항상 상품 감사하다 배송 주문 최고
--------------------------------------
오뚜기 진비빔면156g*4 1번째 토픽 : 공란 매콤 양도 제일 애용 아주 소스 추천 아이 조금
오뚜기 진비빔면156g*4 2번째 토픽 : 공란 먹다 보다 역시 넣다 처음 양념 비비다 살다 골뱅이
오뚜기 진비빔면156g*4 3번째 토픽 : 공란 최고 재다 잇다 가성 나다 진짜 먹다 갈아타다 넘다
오뚜기 진비빔면156g*4 4번째 토픽 : 공란 배송 감사하다 구입 가격 시키다 정말 상품 먹다 사보다
오뚜기 진비빔면156g*4 5번째 토픽 : 공란 구매 자다 자주 제품 항상 받다 사다 먹다 상품
--------------------------------------
농심 찰비빔면 (130g*4입) 1번째 토픽 : 공란 구매 자주 항상 제품 상품 주문 애용 가격 재다
농심 찰비빔면 (13

- 토픽의 갯수를 공통적으로 5개로 군집화 하였으나, 최적화된 내용이 들어있지 않아 내용이 조금 이상한 것으로 판단됨
- 군집화 최적 코드를 작성 -> 실루엣 계수 활용

In [ ]:
from sklearn.metrics import silhouette_score

In [ ]:
def silhouette_point(vectorizer,input_text_list):
  k_range = range(2,11)

  best_n = -1
  best_score = -1
  best_score_list = []
  for k in k_range:

      features = vectorizer.fit_transform(input_text_list)

      from sklearn.decomposition import LatentDirichletAllocation

      lda_model = LatentDirichletAllocation(n_components=k, random_state=55) #인스턴스화 #n_components 토픽의 갯수
      lda_model.fit(features) #교육

      clusters = lda_model.transform(features).argmax(axis=1)  # 각 문서의 클러스터 할당

      score = silhouette_score(features,clusters)
      print('k :',k,'score :',score)

      if score > best_score:
          best_n = k
          best_score = score

  print('best n :', best_n, 'best score :', best_score)
  best_score_list.append(best_n)

  return best_score_list

In [ ]:
best_score = []
counter = 0
for i in dfs:
  # db저장시 발생했던 고유값 제거
  i = i.drop(columns = '_id')

  # 라벨링 함수 적용
  i['textLevel'] = i['textLevel'].apply(labeling)

  # 긍정문 추출
  processing_df = i.query('textLevel == 1')

  # 토픽 추출
  try:
    best_score.append(silhouette_point(vectorizer,processing_df['textContent']))
  except:
    best_score.append(f'오류 발생 : {counter}번재 csv')

  counter += 1

k : 2 score : 0.02835717931804569
k : 3 score : 0.026478923383984623
k : 4 score : 0.042129902677417154
k : 5 score : 0.03242338172741929
k : 6 score : 0.035179676035711795
k : 7 score : 0.03629682236870501
k : 8 score : 0.032276231383014414
k : 9 score : 0.04129249140621835
k : 10 score : 0.026193034033645958
best n : 4 best score : 0.042129902677417154
k : 2 score : 0.022553620027701927
k : 3 score : 0.028458482365526543
k : 4 score : 0.019817163766221335
k : 5 score : 0.01933712909242506
k : 6 score : 0.014776469976684374
k : 7 score : 0.013144102382134383
k : 8 score : 0.01697857455462934
k : 9 score : 0.016099277841067447
k : 10 score : 0.013352797572354334
best n : 3 best score : 0.028458482365526543
k : 2 score : 0.03855138284796299
k : 3 score : 0.041292324696000325
k : 4 score : 0.04007768214412943
k : 5 score : 0.03884803264741444
k : 6 score : 0.023529331184898018
k : 7 score : 0.03188714632898794
k : 8 score : 0.025136879657841407
k : 9 score : 0.02358164917967951
k : 10 sc